cds-plugin.js

```diff
@@ -0,0 +1,125 @@
+const { readdir, stat } = require('node:fs/promises')
+const { normalize } = require('node:path')
+const cds = require('@sap/cds')
+const util = require('util')
+const exec = util.promisify(require('child_process').exec)
+
+const { fs, path } = cds.utils
+const DEBUG = cds.debug('cli|build')
+const BUILD_CONFIG = 'tsconfig.cdsbuild.json'
+
+/**
+ * Check if a tsconfig file exists.
+ */
+const tsConfigExists = () => fs.existsSync('tsconfig.json')
+
+/**
+ * Check if separate tsconfig file that is used for building the project.
+ * @returns {boolean}
+ */
+const buildConfigExists = () => fs.existsSync(BUILD_CONFIG)
+
+/**
+ * @param {string} dir The directory to remove.
+ */
+const rmDirIfExists = dir => {
+    try { fs.rmSync(dir, { recursive: true }) } catch { /* ignore */ }
+}
+
+/**
+ * Remove files with given extensions from a directory recursively.
+ * @param {string} dir The directory to start from.
+ * @param {string[]} exts The extensions to remove.
+ * @returns {Promise<void>}
+ */
+const rmFiles = async (dir, exts) => fs.existsSync(dir) 
+    ? Promise.all(
+        (await readdir(dir))
+            .map(async file => {
+                const filePath = path.join(dir, file)
+                if ((await stat(filePath)).isDirectory()) {
+                    return rmFiles(filePath, exts)
+                } else if (exts.some(ext => file.endsWith(ext))) {
+                    fs.unlinkSync(filePath)
+                }
+            })
+    )
+    : undefined
+
+// requires @sap/cds-dk version >= 7.5.0
+cds.build?.register?.('typescript', class extends cds.build.Plugin {
+    static taskDefaults = { src: '.' }
+    static hasTask() { return tsConfigExists() }
+
+    // lower priority than the nodejs task
+    get priority() { return -1 }
+
+    get #appFolder () { return cds?.env?.folders?.app ?? 'app' }
+
+    get #modelDirectoryName () {
+        try {
+            // expected format: { '#cds-models/*': [ './@cds-models/*/index.ts' ] }
+            //                                       ^^^^^^^^^^^^^^^ 
+            //                             relevant part - may be changed by user
+            const config = JSON.parse(fs.readFileSync ('tsconfig.json', 'utf8'))
+            const alias = config.compilerOptions.paths['#cds-models/*'][0]
+            const directory = alias.match(/(?:\.\/)?(.*)\/\*\/index\.ts/)[1]
+            return normalize(directory)  // could contain forward slashes in tsconfig.json
+        } catch {
+            DEBUG?.('tsconfig.json not found, not parsable, or unconclusive. Using default model directory name')
+        }
+        return '@cds-models'
+    }
+
+    init() {
+        this.task.dest = path.join(cds.root, 'gen', 'srv')
+    }
+
+    async #runCdsTyper () {
+        DEBUG?.('running cds-typer')
+        await exec(`npx @cap-js/cds-typer "*" --outputDirectory ${this.#modelDirectoryName}`)
```

Comment:
Wouldn't it make sense to have a cds-typer API that could be called from both the build plugin and the typer CLI?

`this.model()` returns a compiled CSN representing the correct transitive closure of the model. 
I assume that the typer generates types for all entities and type definitions. 
The corresponding model files and folders are either determined by the cds build using `cds.resolve('*')` or defined by the user as part of the `build task configuration`. 
On the one hand, this would ensure that the same default model scope is used as for other build tasks. On the other hand, users might expect the typer to use the given model options as configured in the build task.

Reply:
Very good point! While I would prefer to just feed `this.model()` into the typer api, this is currently not possible, as cds-typer requires two flavours of CSN (xtended and inferred). So we'll still need to read the model from the files, but we at least got rid of the spawned subprocess. 👍